parity-reconstruct/src/lib.rs

```diff
@@ -0,0 +1,526 @@
+//! Reconstruction algorithm for data from parity data.
+//!
+//! Below, we use N to indicate the number of blocks in the data
+//! L is used to indicate the maximum number of missing data block
+//! that we are prepared to handle
+//!
+//! We use U and V for BitArray backing storages of size sufficient
+//! to handle at least N bits (U) or L bits (V). This can't be done
+//! directly due to limitations in the rust type system.
+#![no_std]
+#![deny(missing_docs)]
+
+pub mod lfdbt;
+
+use bitvec::{array::BitArray, view::BitViewSized};
+
+/// Trait describing a parity matrix.
+pub trait ParityMatrix<U: BitViewSized> {
+    /// Get the mth row of the parity matrix.
+    ///
+    /// All rows should be of exactly length N.
+    ///
+    /// For m < N, this should be a row with exactly
+    /// one true, in position m.
+    ///
+    /// Any excess bits in the bitarray MUST be false
+    fn row(&self, m: usize) -> BitArray<U>;
+}
+
+/// Storage trait for matrix data
+///
+/// Note: the maximum number of missing data blocks that we
+/// can handle is assumed to be BitArray<V>.
+pub trait MatrixStorage<V: BitViewSized> {
+    /// Store a row of the parity reconsturction matrix.
+    ///
+    /// Called at most once for each value of m
+    ///
+    /// When called for m, data has the following guarantees
+    ///  data[m] = true
+    ///  data[i] = false for all i > m
+    fn set_row(&mut self, m: usize, data: BitArray<V>);
+    /// Get a row of the parity reconsturction matrix.
+    ///
+    /// Note: can be called before corresponding set_row call.
+    /// in that case, the result should be all false.
+    fn row(&self, m: usize) -> BitArray<V>;
+}
+
+/// Storage trait for parity blocks
+pub trait ParityStorage<const BLOCKSIZE: usize> {
+    /// Store a parity block.
+    ///
+    /// Called at most once for each value of m.
+    fn store(&mut self, m: usize, data: [u8; BLOCKSIZE]);
+    /// Get a block previously stored
+    ///
+    /// Called only after a corresponding store call
+    /// for m.
+    fn get(&self, m: usize) -> [u8; BLOCKSIZE];
+}
+
+/// Storage trait for reconstructed data.
+pub trait DataStorage<const BLOCKSIZE: usize> {
+    /// Store a newly received or reconstructed block.
+    ///
+    /// Called exactly once for each value of m in 0..N
+    fn store(&mut self, m: usize, data: [u8; BLOCKSIZE]);
+    /// Get a block previously received.
+    ///
+    /// Called only after a corresponding store call
+    /// for m.
+    fn get(&self, m: usize) -> [u8; BLOCKSIZE];
+}
+
+/// Status after handling a new block
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum BlockResult {
+    /// More data is needed before a complete
+    /// reconstruction can be made
+    NeedMore,
+    /// Too many data blocks are still missing
+    /// to accept a parity block.
+    TooManyMissing,
+    /// Sufficient data has been received, the
+    /// data store now contains a full copy of
+    /// the original data.
+    Done,
```

Comment:
While not strictly necessary, it'd be nice if this variant returned how much data we actually have.